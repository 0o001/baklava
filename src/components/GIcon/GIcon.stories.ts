import { storiesOf } from '@storybook/vue';
import GIcon from './GIcon.vue';

storiesOf('GIcon', module)
  .addParameters({ component: GIcon })
  .add('Example', () => ({
    components: { GIcon },
    template: `<div> <GIcon name="alert-circle" strokeWidth='2px' /> <GIcon stroke="black" size="44px" fill="var(--orange-100)" name="star" /> </div>`,
  }))