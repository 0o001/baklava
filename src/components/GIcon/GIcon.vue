<template>
  <svg
    v-bind="this.attrs"
    v-on="$listeners"
    :class="classNames"
    :dataName="this.name"
    :dataTags="tags"
    :dataType="this.name"
    v-html="icon.contents"
  />
</template>

<script lang="ts">

import feather from 'feather-icons';

export default {
  name: 'GIcon',
  props: {
    animation: {
      type: String,
      default: undefined,
    },
    animationSpeed: {
      type: String,
      default: undefined,
    },
    fill: {
      type: String,
      default: 'none',
    },
    size: {
      type: [Number, String],
      default: '24px',
    },
    stroke: {
      type: String,
      default: 'currentColor',
    },
    strokeLinecap: {
      type: String,
      default: 'round',
    },
    strokeLinejoin: {
      type: String,
      default: 'round',
    },
    strokeWidth: {
      type: [Number, String],
      default: 2,
    },
    tag: {
      type: String,
      default: 'i',
    },
    name: {
      type: String,
      required: true,
      validator (name) {
        if (!feather.icons[name]) {
          throw new Error(`"${name}" is not an available icon type.`);
        }
        return true;
      },
    },
    color: {
      type: String,
      default: undefined,
    },
  },
  computed: {
    icon () {
      return feather.icons[this.name];
    },
    tags () {
      return feather.icons[this.name] ? feather.icons[this.name].tags : '';
    },
    classNames () {
      const classes = {
        'g-icon': true,
        feather: true,
        [`feather--${this.name}`]: true,
        [`feather--${this.animation}`]: this.animation,
        [`feather--${this.animationSpeed}`]: this.animationSpeed,
      };
      return classes;
    },
    attrs () {
      return {
        ...this.icon.attrs,
        fill: this.fill,
        height: this.size,
        stroke: this.color ? `var(--${this.color})` : this.stroke,
        'stroke-linecap': this.strokeLinecap,
        'stroke-linejoin': this.strokeLinejoin,
        'stroke-width': this.strokeWidth,
        width: this.size,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
@keyframes feather--spin {
  from {
    transform: rotate(0);
  }
  to {
    transform: rotate(360deg);
  }
}
.feather {
  display: inline-block;
  overflow: hidden;
  &--spin {
    animation: feather--spin 2s linear infinite;
  }
  &--pulse {
    animation: feather--spin 2s infinite steps(8);
  }
  &--slow {
    animation-duration: 3s;
  }
  &--fast {
    animation-duration: 1s;
  }
  &__content {
    display: block;
    height: inherit;
    width: inherit;
  }
}
</style>
