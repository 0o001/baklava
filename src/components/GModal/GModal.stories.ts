import { storiesOf } from '@storybook/vue';
import GModal from '../GModal';
import GButton from '../GButton';
import GBox from '../GBox';
import GText from '../GText';

storiesOf('GModal', module)
  .addParameters({ component: GModal })
  .add('Default', () => ({
    components: { GBox, GModal, GButton, GText },
    template: `
      <GBox style="height: 1000px">
        <GButton @click="showModal = true">Open Modal</GButton>
        <GModal :title="'Custom Title Here'" v-model="showModal">
          <template #body>
            <GText variant="body">Cras mattis consectetur purus sit amet fermentum.</GText>
          </template>
          <template #footer>
            <GButton variant="danger" @click="showModal = false">Close</GButton>
          </template>
        </GModal>
      </GBox>
    `,
    data () {
      return {
        showModal: true,
      };
    },
  })).add('With Size', () => ({
    components: { GBox, GModal, GButton, GText },
    template: `
    <GBox>
    <GButton @click="showModal = true">Open Modal</GButton>
    <GModal large :title="'Custom Title Here'" v-model="showModal">
      <template #body>
        <GText variant="body">Cras mattis consectetur purus sit amet fermentum.</GText>
      </template>
      <template #footer>
        <GButton variant="danger" @click="showModal = false">Close</GButton>
      </template>
    </GModal>
    </GBox>
  `,
    data () {
      return {
        showModal: false,
      };
    },
  })).add('With Scrollable', () => ({
    components: { GBox, GModal, GButton, GText },
    template: `
    <GBox>
    <GButton @click="showModal = true">Open Modal</GButton>
    <GModal scrollable :title="'Custom Title Here'" v-model="showModal">
      <template #body>
        <GText variant="body">Cras mattis consectetur purus sit amet fermentum.</GText>
        <GText v-for="(_, index) in Array(15)" :key="index" class="g-mt-32" variant="body">Cras mattis consectetur purus sit amet fermentum.</GText>
      </template>
      <template #footer>
        <GButton variant="danger" @click="showModal = false">Close</GButton>
      </template>
    </GModal>
    </GBox>
  `,
    data () {
      return {
        showModal: false,
      };
    },
  }));
