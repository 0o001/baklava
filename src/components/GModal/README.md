# GModal

## Slots

| Name     | Description |
| -------- | ----------- |
| `body`   |             |
| `footer` | &nbsp;      |

## Props

| Name                 | Type      | Description | Default |
| -------------------- | --------- | ----------- | ------- |
| `v-model` *required* | `Boolean` |             |         |
| `title`              | `String`  |             |         |
| `large`              | `Boolean` |             | `false` |
| `small`              | `Boolean` |             | `false` |
| `scrollable`         | `Boolean` |             | `false` |

## Computed Properties

| Name            | Type     | Description                                      |
| --------------- | -------- | ------------------------------------------------ |
| `classNames`    | `object` | **Dependencies:** `large`, `small`, `scrollable` |
| `hasFooterSlot` | `any`    | **Dependencies:** `$slots`                       |

## Events

| Name    | Description                                                     |
| ------- | --------------------------------------------------------------- |
| `input` | <br/>**Arguments**<br/><ul><li>**`!this.value: any`**</li></ul> |
| `close` | &nbsp;                                                          |

## Methods

### close()

**Syntax**

```typescript
close(): void
```

