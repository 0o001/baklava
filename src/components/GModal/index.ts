import GModal from './GModal.vue';

export default GModal;
