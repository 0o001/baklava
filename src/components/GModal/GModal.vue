<template>
  <transition name="modal-fade">
    <GBox
      class="g-modal"
      v-show="value"
    >
      <GBox
        class="g-modal-mask"
        @click="close"
      />
      <GBox
        :class="classNames"
        class="g-modal-wrapper"
      >
        <GBox
          :p="30"
          class="g-modal-container g-d-flex g-d-direction-column"
        >
          <GBox
            flex
            justify="space-between"
            class="g-modal-header"
          >
            <GText variant="headline-04">
              {{ title }}
            </GText>
            <GIcon
              name="x"
              color="light-grey-500"
              @click="close"
            />
          </GBox>

          <GBox
            :class="hasFooterSlot ? 'g-my-30' : 'g-mt-30'"
            class="g-modal-body"
          >
            <slot name="body" />
          </GBox>

          <GBox class="g-modal-footer">
            <slot name="footer" />
          </GBox>
        </GBox>
      </GBox>
    </GBox>
  </transition>
</template>

<script lang="ts">
import GBox from '../GBox/GBox.vue';
import GText from '../GText/GText.vue';
import GIcon from '../GIcon/GIcon.vue';

export default {
  name: 'GModal',
  components: { GBox, GText, GIcon },
  props: {
    value: {
      type: Boolean,
      required: true,
    },
    title: {
      type: String,
      required: false,
    },
    large: {
      type: Boolean,
      default: false,
    },
    small: {
      type: Boolean,
      default: false,
    },
    scrollable: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    classNames () {
      return {
        '-large': this.large,
        '-small': this.small,
        '-scrollable': this.scrollable,
      };
    },
    hasFooterSlot () {
      return !!this.$slots.footer;
    },
  },
  methods: {
    close (): void {
      this.$emit('input', !this.value);
      this.$emit('close');
    },
  },

  beforeDestroy () {
    document.body.classList.remove('g-no-scroll');
  },

  watch: {
    value: {
      handler (newValue: boolean) {
        if (newValue) {
          document.body.classList.add('g-no-scroll');
        } else {
          document.body.classList.remove('g-no-scroll');
        }
      },
      immediate: true,
    },
  },
};
</script>
<style lang="scss" scoped>
.g-modal-mask {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 1050;
  width: 100%;
  height: 100%;
  outline: 0;
  display: flex;
  align-items: center;
  overflow-x: hidden;
  overflow-y: auto;
  background-color: rgba(0,0,0,.45);
}

.g-modal-wrapper {
  left: 0;
  right: 0;
  top: 30%;
  width: 100%;
  z-index: 1051;
  margin: 0 auto;
  max-width: 636px;
  position: fixed;
  pointer-events: none;

  &.-large {
    max-width: 858px;
  }

  &.-small {
    max-width: 414px;
  }

  &.-scrollable {
    max-height: calc(100% - 3.5rem);

    .g-modal-container {
      max-height: calc(50vh);
    }

    .g-modal-body {
      overflow-y: auto;
    }
  }

  .g-modal-container {
    position: relative;
    width: 100%;
    pointer-events: auto;
    background-clip: padding-box;
    outline: 0;
    border-radius: 6px;
    box-shadow: 0 30px 30px -20px rgba(39, 49, 66, 0.5);
    background-color: var(--white);

    .g-modal-header {
      svg {
        cursor: pointer;
      }
    }
  }
}

.modal-fade-enter,
.modal-fade-leave-active {
  opacity: 0;
}

.modal-fade-enter-active,
.modal-fade-leave-active {
  transition: opacity .5s ease
}
</style>
