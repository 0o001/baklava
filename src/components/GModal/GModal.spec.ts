import { mount } from '@vue/test-utils';
import GModal from './';
import GIcon from '../GIcon/GIcon.vue';

describe('GModal', () => {
  let wrapper: any;

  const renderWrapper = (options?: any) => {
    wrapper = mount(GModal, {
      propsData: {
        value: false,
      },
      slots: {
        body: '<div>Cras mattis consectetur purus sit amet fermentum.</div>',
        footer: '<button>Close</button>',
      },
      ...options,
    });
  };

  beforeEach(() => {
    renderWrapper();
  });

  it('should match with snapshot', async () => {
    await wrapper.setProps({ value: true, title: 'Custom Title' });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should match with snapshot with size', async () => {
    await wrapper.setProps({ value: true, large: true, scrollable: true, title: 'Custom Title' });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should not render modal component with default values', () => {
    const modalWrapper = wrapper.find('.g-modal').element.style.getPropertyValue('display');

    expect(modalWrapper).toBe('none');
  });

  it('should render modal component with given props', async () => {
    await wrapper.setProps({ value: true, title: 'Custom Title' });

    const modalWrapper = wrapper.find('.g-modal-mask').element.style.getPropertyValue('display');
    const headerText = wrapper.find('.g-modal-header .g-text');

    expect(modalWrapper).toBe('');
    expect(headerText.text()).toBe('Custom Title');
  });

  it('should add conditinial classes', async () => {
    await wrapper.setProps({ value: true, large: true, scrollable: true });

    const modalWrapper = wrapper.find('.g-modal-wrapper');

    expect(modalWrapper.classes()).toContain('-large');
    expect(modalWrapper.classes()).toContain('-scrollable');
  });

  it('should emit close event when click the x button', async () => {
    await wrapper.setProps({ value: true });

    const closeButton = wrapper.findComponent(GIcon);
    closeButton.trigger('click');

    expect(wrapper.emitted().input).toBeTruthy();
  });

  it('should emit close event when click the modal mask', async () => {
    await wrapper.setProps({ value: true });

    const modalMask = wrapper.find('.g-modal-mask');
    modalMask.trigger('click');

    expect(wrapper.emitted().input).toBeTruthy();
  });
});
