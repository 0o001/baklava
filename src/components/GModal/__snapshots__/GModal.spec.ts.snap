// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GModal should match with snapshot 1`] = `
<transition-stub
  name="modal-fade"
>
  <div
    class="g-modal"
    style=""
  >
    <div
      class="g-modal-mask"
    />
     
    <div
      class="g-modal-wrapper"
    >
      <div
        class="g-modal-container g-d-flex g-d-direction-column g-p-30"
      >
        <div
          class="g-modal-header g-d-flex g-d-justify-space-between"
        >
          <div
            class="g-text-main-grey-500 -headline-04 g-text"
          >
            
            Custom Title
          
          </div>
           
          <svg
            class="g-icon feather feather--x"
            dataName="x"
            dataTags="cancel,close,delete,remove,times,clear"
            dataType="x"
            fill="none"
            height="24px"
            stroke="var(--light-grey-500)"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </div>
         
        <div
          class="g-modal-body g-my-30"
        >
          <div>
            Cras mattis consectetur purus sit amet fermentum.
          </div>
        </div>
         
        <div
          class="g-modal-footer"
        >
          <button>
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</transition-stub>
`;

exports[`GModal should match with snapshot with size 1`] = `
<transition-stub
  name="modal-fade"
>
  <div
    class="g-modal"
    style=""
  >
    <div
      class="g-modal-mask"
    />
     
    <div
      class="g-modal-wrapper -large -scrollable"
    >
      <div
        class="g-modal-container g-d-flex g-d-direction-column g-p-30"
      >
        <div
          class="g-modal-header g-d-flex g-d-justify-space-between"
        >
          <div
            class="g-text-main-grey-500 -headline-04 g-text"
          >
            
            Custom Title
          
          </div>
           
          <svg
            class="g-icon feather feather--x"
            dataName="x"
            dataTags="cancel,close,delete,remove,times,clear"
            dataType="x"
            fill="none"
            height="24px"
            stroke="var(--light-grey-500)"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </div>
         
        <div
          class="g-modal-body g-my-30"
        >
          <div>
            Cras mattis consectetur purus sit amet fermentum.
          </div>
        </div>
         
        <div
          class="g-modal-footer"
        >
          <button>
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</transition-stub>
`;
